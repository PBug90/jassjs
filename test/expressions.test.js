const { parse, parserInstance } = require('../lib/JassExpressionParser')

beforeAll(() => {
  parserInstance.performSelfAnalysis()
})

describe('expressions', () => {
  it('addition', () => {
    const input = '1+2'
    expect(parse(input)).toMatchSnapshot()
  })
  it('multiplication', () => {
    const input = '1*2'
    expect(parse(input)).toMatchSnapshot()
  })
  it('division', () => {
    const input = '1/4'
    expect(parse(input)).toMatchSnapshot()
  })
  it('subtraction', () => {
    const input = '1-2'
    expect(parse(input)).toMatchSnapshot()
  })
})
