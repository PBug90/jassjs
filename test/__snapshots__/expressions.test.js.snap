// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expressions addition 1`] = `
Object {
  "children": Object {
    "comparator": Array [
      Object {
        "children": Object {
          "addition": Array [
            Object {
              "children": Object {
                "add": Array [
                  Object {
                    "endColumn": 2,
                    "endLine": 1,
                    "endOffset": 1,
                    "image": "+",
                    "startColumn": 2,
                    "startLine": 1,
                    "startOffset": 1,
                    "tokenType": Object {
                      "CATEGORIES": Array [],
                      "PATTERN": /\\\\\\+/,
                      "categoryMatches": Array [],
                      "categoryMatchesMap": Object {},
                      "isParent": false,
                      "name": "add",
                      "tokenTypeIdx": 29,
                    },
                    "tokenTypeIdx": 29,
                  },
                ],
                "multiplication": Array [
                  Object {
                    "children": Object {
                      "primary": Array [
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 1,
                                "endLine": 1,
                                "endOffset": 0,
                                "image": "1",
                                "startColumn": 1,
                                "startLine": 1,
                                "startOffset": 0,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                  Object {
                    "children": Object {
                      "primary": Array [
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 3,
                                "endLine": 1,
                                "endOffset": 2,
                                "image": "2",
                                "startColumn": 3,
                                "startLine": 1,
                                "startOffset": 2,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                ],
              },
              "name": "addition",
            },
          ],
        },
        "name": "comparator",
      },
    ],
  },
  "name": "expression",
}
`;

exports[`expressions array access 1`] = `
Object {
  "children": Object {
    "comparator": Array [
      Object {
        "children": Object {
          "addition": Array [
            Object {
              "children": Object {
                "multiplication": Array [
                  Object {
                    "children": Object {
                      "primary": Array [
                        Object {
                          "children": Object {
                            "arrayAccess": Array [
                              Object {
                                "children": Object {
                                  "expression": Array [
                                    Object {
                                      "children": Object {
                                        "comparator": Array [
                                          Object {
                                            "children": Object {
                                              "addition": Array [
                                                Object {
                                                  "children": Object {
                                                    "multiplication": Array [
                                                      Object {
                                                        "children": Object {
                                                          "primary": Array [
                                                            Object {
                                                              "children": Object {
                                                                "integer": Array [
                                                                  Object {
                                                                    "endColumn": 12,
                                                                    "endLine": 1,
                                                                    "endOffset": 11,
                                                                    "image": "2",
                                                                    "startColumn": 12,
                                                                    "startLine": 1,
                                                                    "startOffset": 11,
                                                                    "tokenType": Object {
                                                                      "CATEGORIES": Array [],
                                                                      "PATTERN": /\\[0-9\\]\\+/,
                                                                      "categoryMatches": Array [],
                                                                      "categoryMatchesMap": Object {},
                                                                      "isParent": false,
                                                                      "name": "integer",
                                                                      "tokenTypeIdx": 42,
                                                                    },
                                                                    "tokenTypeIdx": 42,
                                                                  },
                                                                ],
                                                              },
                                                              "name": "primary",
                                                            },
                                                          ],
                                                        },
                                                        "name": "multiplication",
                                                      },
                                                    ],
                                                  },
                                                  "name": "addition",
                                                },
                                              ],
                                            },
                                            "name": "comparator",
                                          },
                                        ],
                                      },
                                      "name": "expression",
                                    },
                                  ],
                                  "lsquareparen": Array [
                                    Object {
                                      "endColumn": 11,
                                      "endLine": 1,
                                      "endOffset": 10,
                                      "image": "[",
                                      "startColumn": 11,
                                      "startLine": 1,
                                      "startOffset": 10,
                                      "tokenType": Object {
                                        "CATEGORIES": Array [],
                                        "PATTERN": /\\\\\\[/,
                                        "categoryMatches": Array [],
                                        "categoryMatchesMap": Object {},
                                        "isParent": false,
                                        "name": "lsquareparen",
                                        "tokenTypeIdx": 34,
                                      },
                                      "tokenTypeIdx": 34,
                                    },
                                  ],
                                  "rsquareparen": Array [
                                    Object {
                                      "endColumn": 13,
                                      "endLine": 1,
                                      "endOffset": 12,
                                      "image": "]",
                                      "startColumn": 13,
                                      "startLine": 1,
                                      "startOffset": 12,
                                      "tokenType": Object {
                                        "CATEGORIES": Array [],
                                        "PATTERN": /\\\\\\]/,
                                        "categoryMatches": Array [],
                                        "categoryMatchesMap": Object {},
                                        "isParent": false,
                                        "name": "rsquareparen",
                                        "tokenTypeIdx": 35,
                                      },
                                      "tokenTypeIdx": 35,
                                    },
                                  ],
                                },
                                "name": "arrayAccess",
                              },
                            ],
                            "identifier": Array [
                              Object {
                                "endColumn": 10,
                                "endLine": 1,
                                "endOffset": 9,
                                "image": "arrayident",
                                "startColumn": 1,
                                "startLine": 1,
                                "startOffset": 0,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "identifier",
                                  "tokenTypeIdx": 31,
                                },
                                "tokenTypeIdx": 31,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                ],
              },
              "name": "addition",
            },
          ],
        },
        "name": "comparator",
      },
    ],
  },
  "name": "expression",
}
`;

exports[`expressions division 1`] = `
Object {
  "children": Object {
    "comparator": Array [
      Object {
        "children": Object {
          "addition": Array [
            Object {
              "children": Object {
                "multiplication": Array [
                  Object {
                    "children": Object {
                      "div": Array [
                        Object {
                          "endColumn": 2,
                          "endLine": 1,
                          "endOffset": 1,
                          "image": "/",
                          "startColumn": 2,
                          "startLine": 1,
                          "startOffset": 1,
                          "tokenType": Object {
                            "CATEGORIES": Array [],
                            "PATTERN": /\\\\//,
                            "categoryMatches": Array [],
                            "categoryMatchesMap": Object {},
                            "isParent": false,
                            "name": "div",
                            "tokenTypeIdx": 28,
                          },
                          "tokenTypeIdx": 28,
                        },
                      ],
                      "primary": Array [
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 1,
                                "endLine": 1,
                                "endOffset": 0,
                                "image": "1",
                                "startColumn": 1,
                                "startLine": 1,
                                "startOffset": 0,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 3,
                                "endLine": 1,
                                "endOffset": 2,
                                "image": "4",
                                "startColumn": 3,
                                "startLine": 1,
                                "startOffset": 2,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                ],
              },
              "name": "addition",
            },
          ],
        },
        "name": "comparator",
      },
    ],
  },
  "name": "expression",
}
`;

exports[`expressions multiplication 1`] = `
Object {
  "children": Object {
    "comparator": Array [
      Object {
        "children": Object {
          "addition": Array [
            Object {
              "children": Object {
                "multiplication": Array [
                  Object {
                    "children": Object {
                      "mult": Array [
                        Object {
                          "endColumn": 2,
                          "endLine": 1,
                          "endOffset": 1,
                          "image": "*",
                          "startColumn": 2,
                          "startLine": 1,
                          "startOffset": 1,
                          "tokenType": Object {
                            "CATEGORIES": Array [],
                            "PATTERN": /\\\\\\*/,
                            "categoryMatches": Array [],
                            "categoryMatchesMap": Object {},
                            "isParent": false,
                            "name": "mult",
                            "tokenTypeIdx": 27,
                          },
                          "tokenTypeIdx": 27,
                        },
                      ],
                      "primary": Array [
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 1,
                                "endLine": 1,
                                "endOffset": 0,
                                "image": "1",
                                "startColumn": 1,
                                "startLine": 1,
                                "startOffset": 0,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 3,
                                "endLine": 1,
                                "endOffset": 2,
                                "image": "2",
                                "startColumn": 3,
                                "startLine": 1,
                                "startOffset": 2,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                ],
              },
              "name": "addition",
            },
          ],
        },
        "name": "comparator",
      },
    ],
  },
  "name": "expression",
}
`;

exports[`expressions subtraction 1`] = `
Object {
  "children": Object {
    "comparator": Array [
      Object {
        "children": Object {
          "addition": Array [
            Object {
              "children": Object {
                "multiplication": Array [
                  Object {
                    "children": Object {
                      "primary": Array [
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 1,
                                "endLine": 1,
                                "endOffset": 0,
                                "image": "1",
                                "startColumn": 1,
                                "startLine": 1,
                                "startOffset": 0,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                  Object {
                    "children": Object {
                      "primary": Array [
                        Object {
                          "children": Object {
                            "integer": Array [
                              Object {
                                "endColumn": 3,
                                "endLine": 1,
                                "endOffset": 2,
                                "image": "2",
                                "startColumn": 3,
                                "startLine": 1,
                                "startOffset": 2,
                                "tokenType": Object {
                                  "CATEGORIES": Array [],
                                  "PATTERN": /\\[0-9\\]\\+/,
                                  "categoryMatches": Array [],
                                  "categoryMatchesMap": Object {},
                                  "isParent": false,
                                  "name": "integer",
                                  "tokenTypeIdx": 42,
                                },
                                "tokenTypeIdx": 42,
                              },
                            ],
                          },
                          "name": "primary",
                        },
                      ],
                    },
                    "name": "multiplication",
                  },
                ],
                "sub": Array [
                  Object {
                    "endColumn": 2,
                    "endLine": 1,
                    "endOffset": 1,
                    "image": "-",
                    "startColumn": 2,
                    "startLine": 1,
                    "startOffset": 1,
                    "tokenType": Object {
                      "CATEGORIES": Array [],
                      "PATTERN": /-/,
                      "categoryMatches": Array [],
                      "categoryMatchesMap": Object {},
                      "isParent": false,
                      "name": "sub",
                      "tokenTypeIdx": 30,
                    },
                    "tokenTypeIdx": 30,
                  },
                ],
              },
              "name": "addition",
            },
          ],
        },
        "name": "comparator",
      },
    ],
  },
  "name": "expression",
}
`;
